crates/sui-core/src/authority/authority_store.rs

```diff
@@ -861,58 +835,280 @@ impl AuthorityStore {
             //    (But the lock should exist which means previous transactions finished)
-            // 3. Equivocation possible (different TX) but as long as 2f+1 approves current TX its fine
+            // 3. Equivocation possible (different TX) but as long as 2f+1 approves current TX its
+            //    fine
             // 4. Locks may have existed when we started processing this tx, but could have since
-            //    been deleted by a concurrent tx that finished first. In that case, check if the tx effects exist.
-            let new_locks_to_init: Vec<_> = written
-                .iter()
-                .filter_map(|(_, (object_ref, new_object, _kind))| {
-                    if new_object.is_address_owned() {
-                        Some(*object_ref)
-                    } else {
-                        None
-                    }
-                })
-                .collect();
+            //    been deleted by a concurrent tx that finished first. In that case, check if the
+            //    tx effects exist.
+            self.check_locks_exist(&owned_inputs)?;
+        }
 
-            match update_type {
-                UpdateType::Transaction(_) => {
-                    // After this call completes, new txes can run on the output locks, so all
-                    // output objects must be written already.
-                    self.lock_service
-                        .commit_transaction(transaction_digest, owned_inputs, new_locks_to_init)
-                        .await?;
+        write_batch = self.initialize_locks_impl(write_batch, &new_locks_to_init, false)?;
+        write_batch = self.delete_locks(write_batch, &owned_inputs)?;
+
+        write_batch.write()?;
+        trace!("Finished writing batch");
+
+        Ok(())
+    }
+
+    /// Acquires a lock for a transaction on the given objects if they have all been initialized previously
+    /// to None state.  It is also OK if they have been set to the same transaction.
+    /// The locks are all set to the given transaction digest.
+    /// Returns SuiError::ObjectNotFound if no lock record can be found for one of the objects.
+    /// Returns SuiError::ObjectVersionUnavailableForConsumption if one of the objects is not locked at the given version.
+    /// Returns SuiError::ObjectLockConflict if one of the objects is locked by a different transaction in the same epoch.
+    /// Returns SuiError::ObjectLockedAtFutureEpoch if one of the objects is locked in a future epoch (bug).
+    async fn acquire_transaction_locks(
+        &self,
+        epoch: EpochId,
+        owned_input_objects: &[ObjectRef],
+        tx_digest: TransactionDigest,
+    ) -> SuiResult {
+        // Other writers may be attempting to acquire locks on the same objects, so a mutex is
+        // required.
+        // TODO: replace with optimistic transactions (i.e. set lock to tx if none)
+        let _mutexes = self.acquire_locks(owned_input_objects).await;
```

Comment:
Note: this is one of the very few (maybe only) places in all this code that was modified when moving it over from lock_service.rs